keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,415 @@
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+### MaxSkew
+
+`MaxSkew` and the formula (`matching num - minMatching num < MaxSkew`) is core
+of this KEP, so the exact semantics are clarified as below:
+
+- chicken/egg problem
+
+Let's say we have a 3-zone cluster, and there is no pod in any node yet. Here
+comes a pod, and it wants to be scheduled to a zone which has pods with label
+`foo`. Obviously, there is no qualified node. However, we don't stop here;
+instead, we proceed to check if the incoming pod matches itself on its labels.
+If it does, we would think any node is a fit.
+
+This is actually an existing implication in PodAffinity algorithm. I just want to
+put here again to avoid confusion. And **below examples are all based the
+assumption that incoming pod matches itself on its labels**.
+
+- matching number and min matching number
+
+"matching" number is the number of pods matched on topology domain (defined by
+the global topologyKey). Suppose we have a 3-zone cluster, and there are 3 pods
+in zone1, 2 pods in zone2, 1 pod in zone3. And all pods carry label `foo`:
+
+```
++----------------------------+----------------------------+--------+
+|            zone1           |            zone2           |  zone3 |
++----------------------------+----------------------------+--------+
+| node1a |  node1b  | node1c |  node2a  | node2b | node2c | node3a |
++--------+----------+--------+----------+--------+--------+--------+
+|   pod  | pod, pod |        | pod, pod |        |        |   pod  |
++--------+----------+--------+----------+--------+--------+--------+
+```
+
+Now let's say there comes a pod, it wants to be placed along with pods which
+carries label `foo` in zones.
+
+If global topologyKey is "zone" and maxSkew is "1", then incoming pod can only
+be put into zone3 because for zone1, it violates `matching num (3) - min
+matching num (1) < maxSkey (1). Zone2 violate the formula the same way.
+
+If global topologyKey is "node" and masSkew is "1", things are slightly
+different. Min matching num becomes 0 now, and hence only node1c, node2a and
+node2c are qualified candidates.
+
+- what if a topology domain is infeasible
+
+Suppose we have pods distribution in a 3-zone cluster as 3/3/0, and all pods
+have label `foo`:
+
+```
++-------------+-------------+--------------------+
+|    zone1    |    zone2    | zone3 (infeasible) |
++-------------+-------------+--------------------+
+| pod,pod,pod | pod,pod,pod |                    |
++-------------+-------------+--------------------+
+```
+
+And we have an incoming pod which wants to be scheduled with pods which carry
```

Comment:
I disagree with this paragraph. If zone3 is not feasible for other reasons, scheduling the new pod in any of zone1 and zone2 increases MaxSkew more and therefore, none of these zones should be considered feasible and the new pod should be considered infeasible, unless the spreading is "preferred".

Reply:
Suppose MaxSkew is 1, IMO it doesn't make sense that an incoming pod can land on a "4/3/~0~" cluster, but can't land on a "1/1/~0~" cluster.

Reply:
You point is valid. We need to clearly define the rules in this proposal and rules should be easy to understand for users. For that reason, I would prefer an approach that determines "min matching" by only looking at the topologies matching the "node selector". In that case, a pod will not be feasible in either 4/3/0 or 1/1/0 where the last zone is infeasible. This is simple to understand. If users don't want this, they can go with "preferred" spreading instead of "required".

Reply:
Got it. To be clear, suppose maxSkew is 1: (~zone~ means the zone is infeasible)

- for a "1/1/~0~" cluster, pod can't be placed onto any zone if it's "required"; it can be a fit on zone1/zone2 if it's "preferred"
- for a "2/1/~0~" cluster, pod can't be placed onto any zone if it's "required"; it can be a fit on zone2 if it's "preferred"
- for a "1/1/~1~" cluster, pod can be placed onto zone1 or zone2 for both "required" and "preferred" case
- for a "2/1/~1~" cluster, pod can be placed onto zone2 for both "required" or "preferred" case

Sounds good?

Reply:
All are good except the last one.

> for a "2/1/~1~" cluster, pod can be placed onto zone2 if it's a Predicate; zone1 and zone2 are equally preferred if it's a Priority

Zone2 should still be preferred over zone1 for "preferred" spreading.

Reply:
sry, typo..